doc/developer/design/20210728_error_handling_surfacing_connector_failures.md

```diff
@@ -0,0 +1,189 @@
+# Consistent Error Handling and Surfacing Connector Failures
+
+*I will use connector to refer to either source or sink below. I will use
+`mz_connector` to mean either `mz_source` or `mz_sink` or both.*
+
+## Summary
+
+Currently, Materialize reacts to errors happening at different stages of a
+connectors lifecycle in different ways. For example, an error happening during
+purification/creation of a connector will be reported back to the user
+immediately while errors that happen during runtime are only logged. We should
+agree on what the behavior should be and fix it where needed.
+
+Additionally, I propose to formally add the concept of a *lifecycle* for
+connectors, which will allow the system and users to determine the status of a
+connector.  Right now, this would only be possible by looking through log files
+for error messages and manually associating them with connectors.
+
+## Goals
+
+- unify error handling policies across connectors
+- make it explicit how Materialize should react to errors at different stages
+  of execution
+- introduce connector lifecycle and surface it to users through system tables
+
+## Non-Goals
+
+- introduce more structured error messages, see
+  https://github.com/MaterializeInc/materialize/discussions/5340
+
+## Description
+
+The actual proposal is [way down](#proposed-changes) but it helps to first
+understand the current state.
+
+### Current State
+
+#### Where/when do we interact with external systems?
+
+- When creating a source/sink:
+  - (coord) purification
+- When creating a source/sink or when materialized is restarted:
+  - (coord) initial connector setup, for example:
+  - (coord) reading Kafka consistency topic
+  - (coord) publishing schemas
+  - (coord) creating topics
+  - (coord) creating output file, checking input file
```

Comment:
Listing the keys in an S3 bucket is also a one-time task that can fail.

Reply:
It is, but this is only started after the source was successfully created, I believe. I've looked at this one: https://github.com/MaterializeInc/materialize/blob/480aa30190d7e62ec1fd6843c441739d383db5e2/src/dataflow/src/source/s3.rs#L219

I'm adding these examples to the doc as well, so thanks!